### PR TITLE
ci: fix missing imports, bad func usage, add fail check when release in progress

### PR DESCRIPTION
**Pull request by louis-bompart**
## Proposed changes

- Fixes missing imports that caused https://github.com/coveo/cli/actions/runs/4597808599/jobs/8121171408 :x:
- Fix-it-twice: Typescript checks on the JS files. (I elected a JS check instead of using compilation or ts-node to KISS)
- Add further protection against merging while a release is in progress

## New protections

Currently (assuming it was working), if a PR has all its checks ✅, we as admin could still merge it.
The only solution to prevent an admin to merge is to make a check of the PR to fail
To do that, I used the logic we had for the 'rogue' package-lock protection: a check that exit 1 or 0 depending on the presence of a file or not.
To enforce it, I need to "Requires branches to be up to date before merging", otherwise no further checks would be made.

So now to lock the branch:
 - We verify the main branch hasn't changed
 - We write and commit-push a 'lockfile' into the repo: this is the actual 'lock': from here, until this file get removed from master, nothing can go in or out.
 - We reset the local to the commit before (indeed we don't want to keep this file or commit)
 - We continue the process as usual
 - When we set where `master` should ref, we use 'force' because it is not a fast-forward operation. (i.e. we want to remove 'gitlock'

### CODE REVIEW

**Review by btaillon — APPROVED** [review]

File: utils/release/npm-publish-package.mjs [hunk 1]
```diff
@@ -134,7 +146,8 @@ async function updateWorkspaceDependent(version) {
   )) {
     if (
       dependencies.find(
-        (dependency) => dependency.target === dependencyPackageName
+        (/** @type {{target:string}} **/ dependency) =>
+          dependency.target === dependencyPackageName
```
Comment (mrrajamanickam-coveo) [inline utils/release/npm-publish-package.mjs:150]: Would using typecast `as string` work in this case instead?

File: utils/release/npm-publish-package.mjs [hunk 2]
```diff
@@ -180,15 +204,28 @@ function isPrivatePackage() {
...
+ * @returns {Promise<string>}
+ */
 async function getNextBetaVersion(packageName, nextGoldVersion) {
+  let nextBetaVersion = `${nextGoldVersion}-0`;
   const registryMeta = await fetchNpm(packageName);
+
+  /**
+   * @type {string[]}
+   */
+  // @ts-ignore force-cast.
```
Comment (mrrajamanickam-coveo) [inline utils/release/npm-publish-package.mjs:220]: why is this necessary?

File: utils/release/git-publish-all.mjs [hunk 3]
```diff
@@ -146,6 +154,13 @@ const getCliChangelog = () => {
   });
 })();
 
+/**
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply utils/release/git-publish-all.mjs:157]: It's mostly to 'type' the param of the function, e.g. in typescript it would have been:
```ts
async function commitChanges(releaseNumber: string|number, commitMessage: string, octokit: Octokit) {
```

File: utils/release/npm-publish-package.mjs (hunk 1 above)
Comment (louis-bompart) [reply utils/release/npm-publish-package.mjs:150]: No, because we're in JS.

File: utils/release/npm-publish-package.mjs (hunk 2 above)
Comment (louis-bompart) [reply utils/release/npm-publish-package.mjs:220]: Because `fetchNpm` is loosely typed and return an `unknown` type.

File: utils/release/git-publish-all.mjs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrrajamanickam-coveo) [reply utils/release/git-publish-all.mjs:157]: Yes a brief desc would suffice - it would also make it look less weird / incomplete.

File: utils/release/npm-publish-package.mjs (hunk 1 above)
Comment (mrrajamanickam-coveo) [reply utils/release/npm-publish-package.mjs:150]: Yes right 🤦

File: utils/release/npm-publish-package.mjs (hunk 2 above)
Comment (mrrajamanickam-coveo) [reply utils/release/npm-publish-package.mjs:220]: Got it, thanks. Maybe add a brief note as comment ?

**Review by mrrajamanickam-coveo — APPROVED** [review]

**Comment by louis-bompart** [comment]
Sorry for the re-request @mrrajamanickam-coveo & @btaillon, https://github.com/coveo/cli/pull/1261/commits/c7e4e691deb7a6e2882def7c23d1fc40ddfab163 introduce yet another protection in the PR.

Indeed, without this changeset, we were still able to merge a PR if it passes all the checks.
To prevent that, I copied the logic we have to 'prevent' rogue package-lock.json and adapted it to protect the branch.
Updating pr descr and title with further details

File: utils/release/lock-master.mjs [hunk 4]
```diff
@@ -18,20 +18,37 @@ export const limitWriteAccessToBot = () => changeBranchRestrictions(true);
...
     await octokit.rest.repos.setAdminBranchProtection(COVEO_CLI_MASTER);
   } else {
+    // Allow direct write access to the team
     await octokit.rest.repos.setTeamAccessRestrictions({
       ...COVEO_CLI_MASTER,
       teams: ['dx'],
     });
+    // Allow admins to bypass the status checks
     await octokit.rest.repos.deleteAdminBranchProtection(COVEO_CLI_MASTER);
+    // Do not requires branches to be up to date before merging
```
Comment (mrrajamanickam-coveo) [inline utils/release/lock-master.mjs:48]: Thanks for adding all the inline comments 👍

File: utils/release/git-lock.mjs [hunk 5]
```diff
@@ -29,6 +38,18 @@ const ensureUpToDateBranch = async () => {
   }
 };
 
+/**
+ * This will make .github\workflows\git-lock-fail.yml run and thus fail the associated check.
+ */
+const lockBranch = async () => {
```
Comment (mrrajamanickam-coveo) [inline utils/release/git-lock.mjs:44]: So the purpose of this is to create a failed check that will prevent merging ?

Doesn't github have a way to setting that controls admins from bypassing checks?

Guessing this is not related a concurrency issue that could be solved using [Concurrency Workflow syntax for GitHub Actions - GitHub Docs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency) ?

File: utils/release/npm-publish-package.mjs [hunk 6]
```diff
@@ -170,6 +189,11 @@ function updateDependency(packageJson, dependency, version) {
   }
 }
 
+/**
```
Comment (mrrajamanickam-coveo) [inline utils/release/npm-publish-package.mjs:192]: Couple of functions missing desc here

File: utils/release/git-lock.mjs (hunk 5 above)
Comment (louis-bompart) [reply utils/release/git-lock.mjs:44]: > So the purpose of this is to create a failed check that will prevent merging?

Exactly, because as long as we have to write access to the repo, if we pass the check we will be able to merge 'green' PR.
So I need to paint it red, because if I remove our write access... Well I have no means to reestablish them easily in case something goes wrong.

> Doesn't github have a way to setting that controls admins from bypassing checks?

Yep, and we enable this setting here:
https://github.com/coveo/cli/blob/c0b6610e156efed43cdb80d858333047e8f1b434/utils/release/lock-master.mjs#L38-L39

> Guessing this is not related a concurrency issue that could be solved using [Concurrency Workflow syntax for GitHub Actions - GitHub Docs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency) ?

No indeed, this cannot be resolved by concurrency for several reasons:
 1. If a PR has already passed the checks, it will be able to go through, concurrency wouldn't help prevent that.
 2. If we setup a concurrency constraint on the CI, that means that only one PR can run its checks at a time, this would slow us down
